File: Cpp/src/services/UserService.cc
#include "UserService.h"
#include <iostream>

namespace services {

    UserService::UserService() {}

    models::User UserService::createUser(const std::string& name, const std::string& email) {
        models::User user;
        user.name = name;
        user.email = email;
        int id = repo.create(user);
        user.id = id;
        return user;
    }

    std::optional<models::User> UserService::getUserById(int id) {
        return repo.findById(id);
    }

    std::vector<models::User> UserService::getAllUsers() {
        return repo.findAll();
    }

    std::optional<models::User> UserService::updateUser(int id, const std::string& name, const std::string& email) {
        auto user = repo.findById(id);
        if (!user) return std::nullopt;

        if (!name.empty()) user->name = name;
        if (!email.empty()) user->email = email;

        repo.update(*user);
        return user;
    }

    bool UserService::deleteUser(int id) {
        if (!repo.exists(id)) return false;
        repo.deleteById(id);
        return true;
    }
}

File: JavaSpring/src/main/java/com/magnetaryggdrasil/javaspring/service/UserService.java
package com.magnetaryggdrasil.javaspring.service;

import com.magnetaryggdrasil.javaspring.model.User;
import com.magnetaryggdrasil.javaspring.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class UserService {

    private final UserRepository userRepository;

    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public User createUser(User user) {
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new IllegalArgumentException("Email already exists");
        }
        return userRepository.save(user);
    }

    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }

    public User updateUser(Long id, User userDetails) {
        return userRepository.findById(id).map(user -> {
            if (userDetails.getName() != null) {
                user.setName(userDetails.getName());
            }
            if (userDetails.getEmail() != null) {
                // If email is changing, check for uniqueness
                if (!user.getEmail().equals(userDetails.getEmail()) && userRepository.existsByEmail(userDetails.getEmail())) {
                    throw new IllegalArgumentException("Email already exists");
                }
                user.setEmail(userDetails.getEmail());
            }
            return userRepository.save(user);
        }).orElseThrow(() -> new IllegalArgumentException("User not found with id " + id));
    }

    public void deleteUser(Long id) {
        if (!userRepository.existsById(id)) {
             throw new IllegalArgumentException("User not found with id " + id);
        }
        userRepository.deleteById(id);
    }
}
